packages/labs/compiler/src/lib/ast-fragments.ts

```diff
@@ -7,14 +7,175 @@
 import ts from 'typescript';
 
 export const PartType = {
+  ATTRIBUTE: 1,
   CHILD: 2,
+  PROPERTY: 3,
+  BOOLEAN_ATTRIBUTE: 4,
+  EVENT: 5,
+  ELEMENT: 6,
+  COMMENT_PART: 7,
 } as const;
 
 export type PartType = (typeof PartType)[keyof typeof PartType];
 
-export type TemplatePart = {
-  type: typeof PartType.CHILD;
-  index: number;
+export type TemplatePart =
+  | {
+      type:
+        | typeof PartType.CHILD
+        | typeof PartType.COMMENT_PART
+        | typeof PartType.ELEMENT;
+      index: number;
+    }
+  | {
+      type: typeof PartType.ATTRIBUTE;
+      index: number;
+      name: string;
+      strings: Array<string>;
+      ctorType: PartType;
+    };
+
+const attributePartConstructors = {
+  [PartType.ATTRIBUTE]: 'AttributePart',
+  [PartType.PROPERTY]: 'PropertyPart',
+  [PartType.BOOLEAN_ATTRIBUTE]: 'BooleanAttributePart',
+  [PartType.EVENT]: 'EventPart',
+} as const;
+
+/**
+ * Mapping of part constructors to a unique identifier alias. If undefined, the
+ * constructor will not be imported.
+ */
+export interface AttributePartConstructorAliases {
+  AttributePart?: ts.Identifier;
+  PropertyPart?: ts.Identifier;
+  BooleanAttributePart?: ts.Identifier;
+  EventPart?: ts.Identifier;
+}
+
+/**
+ * An array of names for the attribute constructors that are exported from the
+ * module `"lit-html/private-ssr-support.js"` via the `_$LH` object.
+ */
+const attributePartConstructorNames = Object.values(attributePartConstructors);
+
+/**
+ * Creates the statement:
+ *
+ * ```ts
+ * import * as <namespace> from "<moduleSpecifier>"
+ * ```
+ */
+const createImportNamespaceDeclaration = (
+  factory: ts.NodeFactory,
+  namespace: ts.Identifier,
+  moduleSpecifier: ts.StringLiteral
+) =>
+  factory.createImportDeclaration(
+    undefined,
+    factory.createImportClause(
+      false,
+      undefined,
+      factory.createNamespaceImport(namespace)
+    ),
+    moduleSpecifier,
+    undefined
+  );
+
+/**
+ * Add Parts constructors import. To avoid identifier collisions, each part
+ * constructor is mapped to a unique identifier that must be provided. If no
+ * identifier is provided for a part, then the part will not be emitted.
+ *
+ * This uses a namespace import for g3 compatibility.
```

Comment:
Consider this rewrite:

```suggestion
/**
 * Add an import for the needed part constructors.
 *
 * By the time this function is called, we've already transformed the templates in the 
 * file, and discovered which part constructors we need, and those transforms have 
 * generated unique identifiers for each, so this function's job is to make sure the 
 * requested constructors are available as those identifiers.
 *
 * This uses a namespace import for g3 compatibility.
```